UndertaleModLib/UndertaleLists.cs

```diff
@@ -7,171 +7,214 @@
+        public void Serialize(UndertaleWriter writer)
         {
-            writer.Write((uint)Count);
-            for (int i = 0; i < Count; i++)
+            // Write list count
+            int count = Count;
+            writer.Write((uint)count);
+
+            // Write actual objects
+            int i = 0;
+            try
             {
-                try
-                {
-                    writer.WriteUndertaleObject<T>(this[i]);
-                }
-                catch (UndertaleSerializationException e)
+                for (; i < count; i++)
                 {
-                    throw new UndertaleSerializationException(e.Message + "\nwhile writing item " + (i + 1) + " of " + Count + " in a list of " + typeof(T).FullName, e);
+                    writer.WriteUndertaleObject(this[i]);
                 }
             }
+            catch (UndertaleSerializationException e)
+            {
+                throw new UndertaleSerializationException($"{e.Message}\nwhile writing item {i + 1} of {count} in a list of {typeof(T).FullName}", e);
+            }
         }
 
         /// <inheritdoc />
-        public override void Unserialize(UndertaleReader reader)
+        public void Unserialize(UndertaleReader reader)
         {
+            // Read count and initialize list with that capacity
             uint count = reader.ReadUInt32();
             Clear();
             SetCapacity(count);
-            for (uint i = 0; i < count; i++)
+
+            // Read actual objects
+            uint i = 0;
+            try
             {
-                try
+                for (; i < count; i++)
                 {
                     InternalAdd(reader.ReadUndertaleObject<T>());
                 }
-                catch (UndertaleSerializationException e)
-                {
-                    throw new UndertaleSerializationException(e.Message + "\nwhile reading item " + (i + 1) + " of " + count + " in a list of " + typeof(T).FullName, e);
-                }
+            }
+            catch (UndertaleSerializationException e)
+            {
+                throw new UndertaleSerializationException($"{e.Message}\nwhile reading item {i + 1} of {count} in a list of {typeof(T).FullName}", e);
             }
         }
 
         /// <inheritdoc cref="UndertaleObject.UnserializeChildObjectCount(UndertaleReader)"/>
         public static uint UnserializeChildObjectCount(UndertaleReader reader)
         {
+            // Read base object count
             uint count = reader.ReadUInt32();
             if (count == 0)
+            {
+                // Short-circuit if there's no objects
                 return 0;
+            }
 
-            uint totalCount = 0;
-
-            Type t = typeof(T);
-            if (t.IsAssignableTo(typeof(UndertaleResourceRef)))
+            // If the objects are resource refs, the size is known as 4 bytes
+            if (typeof(T).IsAssignableTo(typeof(UndertaleResourceRef)))
```

Comment:
why remove the variable t here? makes things easier to read imo especially since its usded 5 times

Reply:
Put the variable back in place.